source/modulo_new_core/src/translators/parameter_translators.cpp

```diff
@@ -0,0 +1,225 @@
+#include "modulo_new_core/translators/parameter_translators.hpp"
+
+#include <state_representation/space/cartesian/CartesianState.hpp>
+#include <state_representation/space/cartesian/CartesianPose.hpp>
+#include <state_representation/space/joint/JointState.hpp>
+#include <state_representation/space/joint/JointPositions.hpp>
+
+#include <clproto.h>
+
+using namespace state_representation;
+
+namespace modulo_new_core::translators {
+
+rclcpp::Parameter write_parameter(const std::shared_ptr<state_representation::ParameterInterface>& parameter) {
+  switch (parameter->get_parameter_type()) {
+    case ParameterType::BOOL:
+      return {parameter->get_name(), parameter->get_parameter_value<bool>()};
+    case ParameterType::BOOL_ARRAY:
+      return {parameter->get_name(), parameter->get_parameter_value<std::vector<bool>>()};
+    case ParameterType::INT:
+      return {parameter->get_name(), parameter->get_parameter_value<int>()};
+    case ParameterType::INT_ARRAY:
+      return {parameter->get_name(), parameter->get_parameter_value<std::vector<int>>()};
+    case ParameterType::DOUBLE:
+      return {parameter->get_name(), parameter->get_parameter_value<double>()};
+    case ParameterType::DOUBLE_ARRAY:
+      return {parameter->get_name(), parameter->get_parameter_value<std::vector<double>>()};
+    case ParameterType::STRING:
+      return {parameter->get_name(), parameter->get_parameter_value<std::string>()};
+    case ParameterType::STRING_ARRAY:
+      return {parameter->get_name(), parameter->get_parameter_value<std::vector<std::string>>()};
+    case ParameterType::STATE: {
+      switch (parameter->get_parameter_state_type()) {
+        case StateType::CARTESIAN_STATE:
+          return {parameter->get_name(), clproto::to_json(parameter->get_parameter_value<CartesianState>())};
+        case StateType::CARTESIAN_POSE:
+          return {parameter->get_name(), clproto::to_json(parameter->get_parameter_value<CartesianPose>())};
+        case StateType::JOINT_STATE:
+          return {parameter->get_name(), clproto::to_json(parameter->get_parameter_value<JointState>())};
+        case StateType::JOINT_POSITIONS:
+          return {parameter->get_name(), clproto::to_json(parameter->get_parameter_value<JointPositions>())};
+        default:
+          break;
+      }
+      break;
+    }
+    case ParameterType::VECTOR: {
+      auto eigen_vector = parameter->get_parameter_value<Eigen::VectorXd>();
+      std::vector<double> vec(eigen_vector.data(), eigen_vector.data() + eigen_vector.size());
+      return {parameter->get_name(), vec};
+    }
+    case ParameterType::MATRIX: {
+      auto eigen_matrix = parameter->get_parameter_value<Eigen::MatrixXd>();
+      std::vector<double> vec(eigen_matrix.data(), eigen_matrix.data() + eigen_matrix.size());
+      return {parameter->get_name(), vec};
+    }
+    default:
+      break;
+  }
+  throw InvalidParameterException("Parameter " + parameter->get_name() + " could not be written!");
```

Comment:
yes, better custom exceptions will be nice. The last throw in each translator is the fall-through case, where the type is not recognised or supported. This would rarely occur and can be the most generic one. In the other cases, partly where I've marked TODOs, I would want more specific exceptions.

There will be times where encoding / decoding or value casting will fail, but those will already have their appropriate lower-level exceptions (Decoding Exception or InvalidParameterCast etc.) So for now, I either don't catch them, or catch and rethrow them.

I think I would leave some of the TODOs for now, since we will better understand how exceptions should be thrown or handled once it's integrated with the `add_parameter` and `ParameterMap` of the container base class.